geograph/geograph.py

```diff
@@ -892,8 +974,32 @@ def _add_node(
         )
 
     # pylint: disable=dangerous-default-value
-    def _add_nodes(self, node_ids, adjacencies, node_data={}, **data):
-        raise NotImplementedError
+    def _add_nodes(
+        self,
+        node_ids,
+        adjacencies: Dict[int, Iterable[int]],
+        node_data: Dict[int, Dict],
+        requires_sorting: bool = True,
+    ) -> None:
+        """
+        Add multiple nodes to the graph.
+
+        Args:
+            node_ids (Iterable[int]): The ids of the nodes to add.
+            adjacencies (Dict[int, Iterable[int]): Dict mapping node ids to
+                Iterables of node ids which are adjacent to each new node.
+            node_data (Dict[int, Dict]): Dict mapping node ids to dictionaries of
+                node attributes.
+            requires_sorting (bool, optional): Whether or not to sort the dataframe
+                index after adding the nodes. Defaults to True.
+        """
+        for node_id in node_ids:
```

Comment:
```suggestion
        for node_id in node_ids:  # TODO: Speed up by turning into bulk operation
```
We should be able to significantly speed this up by turning it into a bulk operation, at least on the dataframe. Shall we add an issue for that?

Reply:
Good idea!